### PR TITLE
[SPARK-39731][SQL] Fix issue in CSV and JSON data sources when parsing dates in "yyyyMMdd" format with CORRECTED time parser policy

### PR DESCRIPTION
**Pull request by sadikovi**
### What changes were proposed in this pull request?

This PR fixes a correctness issue when reading a CSV or a JSON file with dates in "yyyyMMdd" format: 
```
name,mydate
1,2020011
2,20201203
```
or 
```
{"date": "2020011"}
{"date": "20201203"} 
```

Prior to https://github.com/apache/spark/pull/32959, reading this CSV file would return:
```
+----+--------------+
|name|mydate        |
+----+--------------+
|1   |null          |
|2   |2020-12-03    |
+----+--------------+
```

However, after the patch, the invalid date is parsed because of the much more lenient parsing in `DateTimeUtils.stringToDate`, the method treats `2020011` as a full year:
```
+----+--------------+
|name|mydate        |
+----+--------------+
|1   |+2020011-01-01|
|2   |2020-12-03    |
+----+--------------+
``` 
Similar result would be observed in JSON.

This PR attempts to address correctness issue by introducing a new configuration option `enableDateTimeParsingFallback` which allows to enable/disable the backward compatible parsing. 

Currently, by default we will fall back to the backward compatible behavior only if parser policy is legacy and no custom pattern was set (this is defined in `UnivocityParser` and `JacksonParser` for csv and json respectively).

### Why are the changes needed?
Fixes a correctness issue in Spark 3.4.

### Does this PR introduce _any_ user-facing change?

In order to avoid correctness issues when reading CSV or JSON files with a custom pattern, a new configuration option `enableDateTimeParsingFallback` has been added to control whether or not the code would fall back to the backward compatible behavior of parsing dates and timestamps in CSV and JSON data sources.
- If the config is enabled and the date cannot be parsed, we will fall back to `DateTimeUtils.stringToDate`.
- If the config is enabled and the timestamp cannot be parsed, `DateTimeUtils.stringToTimestamp` will be used.
- Otherwise, depending on the parser policy and a custom pattern, the value will be parsed as null.

### How was this patch tested?

I added unit tests for CSV and JSON to verify the fix and the config option.

### CODE REVIEW

**Comment by sadikovi** [comment]
cc @HyukjinKwon @cloud-fan.

**Comment by HyukjinKwon** [comment]
@Jonathancui123 since you took a look around here lately. Mind reviewing this when you find some time?

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (Jonathancui123) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:233]: This change makes sense to me. Thanks Ivan :))

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala
Comment (Jonathancui123) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:2828]: For completeness, would you consider adding a check for `LEGACY_TIME_PARSER_POLICY` = `EXCEPTION`? Similar to the following?

https://github.com/apache/spark/blob/1193ce78d3efcbe1395305b4b7deb0a195fa09d9/sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala#L2598-L2601

**Review by Jonathancui123 — APPROVED** [review]
LGTM - we don't want to use the legacy parser because it will allow bad data when a custom format is used. However, there are some failing tests that should be addressed: 

```[info] - SPARK-36536: use casting when datetime pattern is not set *** FAILED *** (289 milliseconds)```
```[info] - SPARK-30960: parse date/timestamp string with legacy format *** FAILED *** (73 milliseconds)```
```[info] - SPARK-36536: use casting when datetime pattern is not set *** FAILED *** (289 milliseconds)```

**Review by kamcheungting-db — APPROVED** [review]

**Comment by cloud-fan** [comment]
If the legacy behavior is unreasonable, I think we don't have to keep it. If datetime patten is specified, we should not fall back to the legacy code path, even if it only supports 4 digits like Spark 2.x.

**Comment by sadikovi** [comment]
I actually found a similar issue in JSON data source. I will also address it in this PR and will update the title and description:

test.json
```
{"date": "2020011"}
{"date": "20201203"}
```

```scala
val df = spark.read.schema("date date").option("dateFormat", "yyyyMMdd").json("file:/tmp/test.json")
df.show(false)
```
returns
```
+--------------+
|date          |
+--------------+
|+2020011-01-01|
|2020-12-03    |
+--------------+
```
but before the fix it used to show: 
```
+----------+
|date      |
+----------+
|7500-08-09|
|2020-12-03|
+----------+
```

which is strange either way.

**Review by cloud-fan — APPROVED** [review]

**Comment by sadikovi** [comment]
@Jonathancui123 @kamcheungting-db Could you review this PR again? Thanks.

**Comment by sadikovi** [comment]
Thanks @cloud-fan!

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala
Comment (Jonathancui123) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:369]: Is this technically a breaking change for users who could previously specify an invalid pattern without LEGACY mode? 

Before -- ignore the pattern and parse with `DateTimeUtils.stringToTimestamp`
Now -- it throws an error

We don't support invalid patterns but as a user I would be unhappy to see my code break. I'm unsure if this is actually considered a breaking change because this is such an edge case and the user is already doing something invalid. I'm curious to hear your thoughts.

**Review by Jonathancui123 — APPROVED** [review]
Changes look good to me! I have one question about whether this is a breaking change. Thanks for working on the fix :))

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala
Comment (sadikovi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:369]: This is a good point. It would be a breaking change for users if they were relying on the compatibility fallback.
There could an alternative fix, maybe we can look into updating `DateTimeUtils.stringToDate` but I am not sure.

I can also add a feature flag to control this behaviour in JSON and CSV connectors so users can always opt in to use legacy behaviour. For example, I can a data source option "useLegacyParsing" or something similar. The option could be disabled by default, the exception would contain a message saying that you can enable the option to maintain the previous behaviour. Maybe this could be a good solution.

Let me know if something like that could work, thanks.

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:369]: I think the safest option is to copy-paste the old code of `stringToDate` before https://github.com/apache/spark/pull/32959 and use it here, but that's really ugly and hard to maintain.

I'd like to understand more about the invalid pattern behavior. Will we trigger the fallback for every input row? That sounds like a big perf problem...

Comment (sadikovi) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/csv/UnivocityParserSuite.scala:369]: With the invalid pattern and before this PR, yes, the fallback code would be triggered on every pattern mismatch. With the change, we will just throw an exception parsing those values as nulls. Yes, it does sound like a performance issue but it has been there for some time. 

I agree with copy-paste of stringToDate, I proposed to add a data source config to keep the old behaviour. What do you think?

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala
Comment (sadikovi) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:2828]: Done!

**Comment by sadikovi** [comment]
@Jonathancui123 @cloud-fan I decided to introduce a new config option for JSON and CSV to control this parsing behaviour and have updated the code accordingly. 

Now, users will have a way to enable backward compatible behaviour if they rely on it. The default is the same as with the initial changes: we will not parse the date/timestamp again if time parser policy is not legacy and the custom pattern is set.

**Comment by sadikovi** [comment]
cc @MaxGekk for a review as you are very familiar with JSON and CSV and date/time parsing 🙂.

**Review by cloud-fan — APPROVED** [review]

**Review by Jonathancui123 — APPROVED** [review]
New config makes sense to me - LGTM

**Comment by sadikovi** [comment]
Thanks for the review. Is there anything else I need to do before we can merge this PR? Any additional doc changes? Thanks.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala:116]: Don't we should refer to internal CSV parser from JSON options. Could you rewrite this, please.

**Review by MaxGekk — COMMENTED** [review]
LGTM in general

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala
Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JSONOptions.scala]: I have updated the comment, thanks.

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
@sadikovi Could you re-trigger tests, please.

**Comment by sadikovi** [comment]
Yes, sure. Let me do that.

**Comment by sadikovi** [comment]
@Jonathancui123 Can you review https://github.com/apache/spark/pull/37147/commits/a447b0895cb961a69e9845e0e2fd63bfcebc98fd? The tests were failing so I disabled the flag but maybe we need to revisit the test and/or behaviour of "inferDate" (or confirm it). I am happy to sync offline on this. Thanks.

**Comment by Jonathancui123** [comment]
> @Jonathancui123 Can you review https://github.com/apache/spark/commit/a447b0895cb961a69e9845e0e2fd63bfcebc98fd? 
I've reviewed the changes and written my thoughts [here](https://github.com/apache/spark/pull/36871#issuecomment-1192923903). Depending on our target behavior, we will need to handle these tests differently

**Comment by sadikovi** [comment]
@Jonathancui123 I am going to fix the PR to work with inferDate option and will revert the test changes.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (sadikovi) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:238]: This reorganisation is required to define priority of parsing timestamps as dates when `inferDate = true`. Since this is optional, we try this first and then fall back to the legacy behaviour if it is enabled.

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:238]: cc @cloud-fan @Jonathancui123

Comment (Jonathancui123) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:238]: Thanks for the fix! It also makes sense that you have reverted "fix tests for SPARK-39469". 

This implements @cloud-fan's suggestion and the "Option A" that I described [in my comment](https://github.com/apache/spark/pull/36871#issuecomment-1192923903) which considers type coercion between Dates and Timestamps. I'm okay with this - as a user, it makes sense that when I can turn on `inferDate`, a date entry is detected in a timestamp column and then cast to a timestamp. 

However, I think @HyukjinKwon should take a look. He [said that it would be more future-proof](https://github.com/apache/spark/pull/36871#issuecomment-1193016381) to not provide this functionality.

Comment (sadikovi) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:238]: Thanks 👍. Yeah, it would be good if Hyukjin could take another look.

File: sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:2912]: sorry I'm a bit confused. Why date parsing fails? `2020-01-01` is a valid date string.

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/csv/CSVSuite.scala:2912]: ah, because the format pattern is given but invalid.

**Review by cloud-fan — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/csv/UnivocityParser.scala:238]: It's a small issue within the fallback branch and is probably not an issue without this PR that tries to fix the fallback branch. I'm OK to not open a new PR for it.

**Comment by sadikovi** [comment]
I have re-triggered the failed jobs.

**Comment by sadikovi** [comment]
@HyukjinKwon Do you know what could be going wrong with regard to the test failures? I don't quite understand how to figure out why the tests failed. For example, I get the following error for TPC-DS job: 
```
[info] - q69 (6 seconds, 699 milliseconds)
[info] - q70 (7 seconds, 678 milliseconds)
[info] - q71 (5 seconds, 279 milliseconds)
/home/runner/work/spark/spark/build/sbt-launch-lib.bash: line 77:  2375 Killed                  "$@"
Error: Process completed with exit code 137.
```

**Comment by HyukjinKwon** [comment]
TPC-DS build can be ignored I believe.

**Comment by sadikovi** [comment]
@cloud-fan @HyukjinKwon the tests passed. If there are no reviews, what else is needed to get this PR merged?

**Comment by HyukjinKwon** [comment]
I'll defer to @cloud-fan

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**
